zcash_client_sqlite/src/wallet.rs

```diff
@@ -915,9 +916,46 @@ fn subtree_scan_progress<P: consensus::Parameters>(
         )?;
         Ok(Progress { scan, recover })
     } else {
+        // In case we didn't have information about the tree size at the recover-until
+        // height, get the tree size from a nearby subtree. It's fine for this to be
+        // approximate; it just shifts the boundary between scan and recover progress.
+        let mut get_tree_size_near = |as_of: BlockHeight| {
+            stmt_start_tree_size
+                .query_row(
+                    named_params![":start_height": u32::from(birthday_height)],
+                    |row| row.get::<_, Option<u64>>(0),
+                )
+                .optional()
+                .map(|opt| opt.flatten())
+                .transpose()
+                .or_else(|| {
+                    conn.query_row(
+                        &format!(
+                            "SELECT MIN(shard_index)
+                    FROM {table_prefix}_tree_shards
+                    WHERE subtree_end_height >= :start_height
+                    OR subtree_end_height IS NULL",
+                        ),
+                        named_params! {
+                            ":start_height": u32::from(as_of),
+                        },
+                        |row| {
+                            let min_tree_size = row
+                                .get::<_, Option<u64>>(0)?
+                                .map(|min_idx| min_idx << shard_height);
+                            Ok(min_tree_size)
+                        },
+                    )
+                    .optional()
+                    .map(|opt| opt.flatten())
+                    .transpose()
+                })
+                .transpose()
+        };
```

Comment:
I personally prefer it this way:
```suggestion
        let mut get_tree_size_near = |as_of: BlockHeight| -> Result<_, SqliteClientError> {
            if let Some(tree_size) = stmt_start_tree_size
                .query_row(
                    named_params![":start_height": u32::from(birthday_height)],
                    |row| row.get::<_, Option<u64>>(0),
                )
                .optional()?
                .flatten()
            {
                return Ok(Some(tree_size));
            };
            Ok(conn
                .query_row(
                    &format!(
                        "SELECT MIN(shard_index)
                        FROM {table_prefix}_tree_shards
                        WHERE subtree_end_height >= :start_height
                        OR subtree_end_height IS NULL",
                    ),
                    named_params! {
                        ":start_height": u32::from(as_of),
                    },
                    |row| {
                        let min_tree_size = row
                            .get::<_, Option<u64>>(0)?
                            .map(|min_idx| min_idx << shard_height);
                        Ok(min_tree_size)
                    },
                )
                .optional()?
                .flatten())
        };
```

Reply:
I made the same comment to @nuttycom in mob today, that for this kind of highly-nested situation I find imperative code with early returns (`?`) clearer than functional code with type transposing (despite my general tendency to implement the latter, because in many cases it is great).